Documentation/RepoCompose.md

```diff
@@ -0,0 +1,289 @@
+
+Going forward repos will be given contracts, or APIs, that provide structured data about their dependencies and outputs.  This will give us the necessary data to compose our repos in our growing set of scenarios:
+
+- Linux distro builds from source
+- Official Windows signed build
+- Builds to quickly test and deploy new packages across repos
+
+These contracts will be included in the `run` command [specification](https://github.com/dotnet/buildtools/blob/master/Documentation/RunCommand.md).  This document in particular will be discussing the commands which allow us to compose the outputs of our repos.
+
+## Commands
+
+This document describes the commands used to compose different repos together.  There is a larger set of commands including `build`, `sync` and `clean` which is [described separately](https://github.com/dotnet/buildtools/blob/master/Documentation/Dev-workflow.md). 
+
+### consumes
+
+The `consumes` command returns json output that describes all of the external artifacts consumed by this repo.  This includes NuGet feeds, packages and arbitrary files from the web or file system.  The format of all these items is describe in the Artifact Specification section below. 
+
+The artifacts are grouped into sections: 
+
+- Build Dependencies: artifacts which are referenced by the build output and include NuGet packages, MSI, etc ...  In order to support a composed build these are further divided into floating and static dependencies:
+    - Floating: dependencies where versions can change as a part of a composed build via the `change` command.  This commonly includes CoreFx, CoreClr, etc ...  
+    - Static: dependencies whose version do not change via the `change` command.  This is commonly used for SDK binaries, legacy tools, etc ...  
+- Toolset Dependencies: artifacts used in the production of build output.  This commonly includes NuGet.exe, compilers, etc ... 
+
+These sections are identified respectively by the following JSON sections:
+
+``` json
+"dependencies": {
+    "floating": { 
+        // Build artifacts
+    },
+    "static": { 
+        // Static artifacts
+    },
+    "toolset": { 
+        // Toolset artifacts
+    }
+}
+```
+
+The data in the output is further grouped by operating system.  Builds of the same repo on different operating system can reasonably consume a different set of resources.  The output of `consumes` reflect this and allows per operating system dependencies:
+
+``` json
+{
+    "os-windows": {
+        "dependencies": { } 
+    },
+    "os-linux": {
+        "dependencies": { }
+    }
+}
+```
+
+In the case the `consumes` output doesn't want to take advantage of OS specific dependencies it can specify `"os-all"` as a catch all. 
+
+In addition to artifacts the consume feed can also optionally list any machine prerequitsites needed to build or test the repo:
```

Comment:
At this point no.  Most of our thinking around prereqs is been in the other direction.  Essentially:

> How can we get prereqs down to the absolute minimal possible set? 

Prereqs are a pain because it adds a lot of complexity to creating build environments.  Every prereq is a new installation step on build machines used for composed builds.  This is both a pain to setup and hard to maintain.  

Think back to the days where building VS components relied on having specific builds of VS installed.  Absolute nightmare to keep build machines up to date with that prereq.  Also no real solution if different repos depended on different builds of VS. 

For the time being we're trying to get all of our prereqs into a known state.  From there see what we can get rid of and over time widdle it down to the minimal possible set.  Once get there we can establish a build environment with only those pre-reqs installed and use it to verify that we don't slip up any.